langchain/retrievers/multi_query.py

```diff
@@ -0,0 +1,146 @@
+from typing import List
+
+from pydantic import BaseModel, Field
+
+from langchain import LLMChain
+from langchain.llms import BaseLLM
+from langchain.output_parsers import PydanticOutputParser
+from langchain.prompts import PromptTemplate
+from langchain.schema import BaseRetriever, Document
+
+
+class LineList(BaseModel):
+    lines: List[str] = Field(description="Lines of text")
+
+
+class LineListOutputParser(PydanticOutputParser):
+    def __init__(self) -> None:
+        super().__init__(pydantic_object=LineList)
+
+    def parse(self, text: str) -> LineList:
+        lines = text.strip().split("\n")
+        return LineList(lines=lines)
+
+
+class MultiQueryRetriever:
+
+    """Given a user query, use an LLM to write a set of `num_queries` queries.
+    Retrieve docs for each query. Rake the unique union of all retrieved docs."""
+
+    # Prompt
+    DEFAULT_QUERY_PROMPT = PromptTemplate(
+        input_variables=["question", "num_queries"],
+        template="""You are an AI language model assistant. Your task is 
+        to generate {num_queries} different versions of the given user 
+        question to retrieve relevant documents from a vector  database. 
+        By generating multiple perspectives on the user question, 
+        your goal is to help the user overcome some of the limitations 
+        of distance-based similarity search. Provide these alternative 
+        questions seperated by newlines. Original question: {question}""",
+    )
+
+    def __init__(
+        self,
+        retriever: BaseRetriever,
+        num_queries: int,
+        llm_chain: LLMChain,
+        verbose: bool = True,
+    ) -> None:
+        """Initialize MultiQueryRetriever.
+
+        Args:
+            retriever: retriever to query documents from
+            num_queries: number of queries for the LLM to generate
+            llm_chain: llm_chain for query generation
+            verbose: show the queries that we generated to the user
+
+        Returns:
+            MultiQueryRetriever
+        """
+        self.retriever = retriever
+        self.num_queries = num_queries
+        self.llm_chain = llm_chain
+        self.verbose = verbose
+
+    @classmethod
+    def from_llm(
+        cls, retriever: BaseRetriever, num_queries: int, llm: BaseLLM
+    ) -> "MultiQueryRetriever":
+        """Initialize from llm using default template.
+
+        Args:
+            retriever: retriever to query documents from
+            num_queries: number of queries for the LLM to generate
+            llm: llm for query generation using DEFAULT_QUERY_PROMPT
+
+        Returns:
+            MultiQueryRetriever
+        """
+        llm_chain = LLMChain(llm=llm, prompt=cls.DEFAULT_QUERY_PROMPT)
```

Comment:
we should expose this so the user can pass it in

eg have DEFAULT_QUERY_PROMPT outside of the class, and then do
```
from_llm(..., prompt=DEFAULT_QUERY_PROMPT)
```